File: typescript/15. 3Sum.ts
function threeSum(nums: number[]): number[][] {
  const target = 0;
  const result = new Array();

  nums = nums.sort((a, b) => a - b);

  for (let i = 0; i < nums.length - 2; i++) {
    if (nums[i] > target) break;

    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let l = i + 1;
    let r = nums.length - 1;

    while (l < r) {
      const sum = nums[i] + nums[l] + nums[r];

      if (sum === target) {
        result.push([nums[i], nums[l], nums[r]]);

        while (nums[l] === nums[l + 1]) l++;
        while (nums[r] === nums[r - 1]) r--;

        l++;
        r++;
      } else if (sum < target) {
        l++;
      } else {
        r--;
      }
    }
  }

  return result;
}

File: golang/daily/15. 3Sum.go
package daily

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)

	var result [][]int
	for i := 0; i < len(nums)-2; i++ {
		if i != 0 && nums[i] == nums[i-1] {
			continue
		}
		target := nums[i] * -1

		l := i + 1
		r := len(nums) - 1

		for l < r {
			sum := nums[l] + nums[r]
			if sum > target {
				r--
			} else if sum < target {
				l++
			} else {
				result = append(result, []int{nums[i], nums[l], nums[r]})
				r--
				l++

				for nums[l] == nums[l-1] && l < r {
					l++
				}

				for nums[r] == nums[r+1] && l < r {
					r--
				}
			}
		}
	}
	return result
}
